### ISSUE TITLE
HasValue interface oblivious to Bindings creates problems

### ISSUE THREAD

**Issue by pwilkin** [vaadin/framework#9949]
The new data model in Vaadin 8 has simplified a lot of things and got rid of underlying Properties/Items/Containers. However, the new API has arguably gone too far in simplifying the data binding mechanism. 

As of now, `HasValue`s are completely oblivious to the fact that any bindings have been performed on them. This raises a few important problems: 
* if your form ever processes data in the background (fields with dependencies / recalculations, for example) and you want to `discard()`. the data, you cannot do this from the field as it has no `discard()` method anymore - you have to `setValue()` manually
* since bindings are handled via value change listeners and there is no way to pick out the specific listener responsible for the binding, one cannot effectively use `setValue()` as a `discard()` substitute
* the only way to do a `discard()` effectively requires having access to the binder and doing a somewhat counterintuitive `setBean(getBean())`. However, this forces a reload of all bound fields
* if you have a `ValueChangeListener` that depends on the underlying data object being updated, then managing this is a nightmare, because `initFieldValue()` causes the data binding listener to be removed and then readded to the end of the queue, making it trigger at the end of the listener list - so, effectively, all other listeners are perfomed without the value actually being written to the data source yet

To summarize, the current binding mechanism is a nightmare for anyone that ever does background processing of forms and has some form logic such as "change value in one field -> value in other field changes -> form has to reload to incorporate the change".

My proposal is the following: 
* make `HasValue` expose `getBinders()` and make bindings actually register the binder
* make the `ValueChangeListener` created by the binder privileged by adding it always to the start of the listener queue (might require some sort of `addValueChangeListenerFirst()` method
* expose `discard()` for `HasValue` again via the binders (possibly make `initFieldValue` public?), maybe also expose an `isBound()` method
* make `discardAll()` in a `Binder` as a shortcut for the counterintuitive `setBean(getBean())`

This would be of huge help for anyone that uses dynamic and not static forms in their projects (there have been complaints already about problems with changing validations within the new data model).

**Comment by pwilkin** [comment]
I'd be perfectly willing to make a pull request to master that performs those changes, however, the question is whether they would get accepted - is there any reason **not** to do the abovementioned things?

**Comment by Legioth** [comment]
There is one big reason to not fix the described issues in the suggested way: it would add complexity to every `HasValue` implementation. One of the goals with the new data API has been to simplify the implementation of field components compared to the ~2000 lines of complexity that were necessary in `AbstractField` from version 7.

We want that each individual field implementation can be as "dumb" as possible, and anything that is common to all field types should instead be handled through `Binder`.

Ensuring that a bound bean has already been updated when a `ValueChangeListener` is notified can be done by registering the listener using `Binder.addValueChangeListener(ValueChangeListener<?> listener)` instead of adding it directly to the field. One small drawback with this approach is that your listener would have to filter out events that originate from fields that it's not interested in. To help deal with this, we also consider adding a similar helper to `Binding`.

To help deal with the same situation for which you have suggested `HasValue.discard()`, we would instead suggest adding `Binding.readFromBean(BEAN bean)`, and correspondingly also `Binding.writeToBean(BEAN bean)` for the opposite case. These methods would also be useful for many other situations. To help using these, we would also make it easier to find all bindings used by a `Binder` and to find the binding that corresponds to a specific field.

What do you think about this approach?

**Comment by pwilkin** [comment]
To be honest, I don't see a collision between what you're saying and what I'm proposing. 

I'm not saying that all the functionalities should be moved back into `HasValue`. The only extension to `HasValue` that I'm postulating is that the `HasValue` becomes aware of its binders, which makes it much easier to handle dynamic forms. The way I see it, it's just a couple lines of code. All the other matters basically amount to exposing more methods within the `Binder` interface and the `Binding` interface and also allowing for more flexibility in i.e. modifying a `Binding`s properties. I don't see a problem with using `Binding.readFromBean(...)` instead of `HasValue.discard()` (as long as it allows you to reread just one field's value and not all the `Binding`s from a given `Binder` as it is now). 

The way I see it, adding `Binder` registration would be a big convenience, but the killers at the moment are the hidden interfaces for `Binding` and the inability to do anything with `Binding`s once they're established.

**Comment by pwilkin** [comment]
Oh, and adding that listener to the `Binding` would certainly be of huge help in making sure that I'm handling a value change after it has been passed to the data source. However, that's also another place where the `Binder` registration would be convenient (or maybe even a `Binding` registration?), so that I could do something like `field.getBindings().forEach(x -> x.addValueChangeListener(listener))`.

**Comment by Legioth** [comment]
> The way I see it, it's just a couple lines of code.

It's a couple of lines of code in *every* `HasValue` implementation out there. Some implementations might get the change "for free" through extending `AbstractField`, but others will not. This means that adding this one method would significantly break backwards compatibility.

Another aspect is separation of concerns: The purpose of a field component is to show a value in the browser, let the user modify that value and fire events when the value changes. Why should it make a distinction between different types of listeners for the same event?

For these reasons, I would see addition something like `HasValue.getBindings()` as a last resort in case there are no other ways of supporting some specific use case. I still have a hard time understanding any specific case it is that would benefit from this functionality.

> do something like `field.getBindings().forEach(x -> x.addValueChangeListener(listener))`.

Some open questions about this example:
* Why would the same field be managed by multiple bindings? Sounds like a really confusing case. Which binding is authoritative?
* If there is application code that is concerned about bindings, then wouldn't it make more sense to use the binding as the primary object that is passed around instead of passing the field?

**Comment by pwilkin** [comment]
You are probably right about the separation of concerns, however, as of now it is impossible to do a proper separation of concerns because the `addValueChangeListener` for the `Binding` (mentioned in the discussion above). Therefore, I have to keep track of value change listeners and make sure that mine is always registered last so that the value change is passed to the data source. 

As for "field managed by multiple bindings" - as long as it's possible to bind one field to multiple data sources, this should be supported simply because it's possible. However, you are right about the fact that it's confusing (certainly breaks `discard()` logic). 

To summarize, you're probably correct in that the proposed changes are largely due to the attachment to the previous API. What would be needed in order to realize the required features in the new API is: 

* allowing a `ValueChangeListener` within a `Binding` to mimic the Vaadin 7 listener behavior where the listener had a guarantee that the value was already passed to the data source (a `ValueChangeListener` on the entire `Binder` is too coarse-grained)
* selecting a `Binding` from a `Binder` based on the `Field` and not just the `propertyId`
* the abovementioned readFromBean(...) and writeFromBean(...) methods for `Binding`
* possibly access to `Binder` from a `Binding`, so that you don't have to keep an extra reference to the `Binder` in addition to any map you might have containing the `Binding`s. Plus, this makes it easy to implement `discard()` and `commit()` as shortcuts for `readFromBean(getBinder().getBean())` and `writeToBean(getBinder().getBean())` respectively
* a convenient API for rebinding (as for now, there is no `unbind()` method and no `rebind(...)` method, so the only way to rebind things is to scrap all listeners on all bound fields and rebuild the entire `Binder`, something that can be horribly inefficient on large forms). Seeing as the `Binder` implementation remembers the `Registration` anyways, doing `unbind(Binding binding)` is probably as simple as: 

```
public void unbind(Binding binding) {
    if (bindings.contains(binding)) {
        bindings.remove(binding);
        if (boundProperties.values().contains(binding)) {
            boundProperties.values().remove(binding);
        }
        if (binding.onValueChange != null) {
            binding.onValueChange.remove();
        }
    }
}
```

What do you think about such as solution?

**Comment by tsuoanttila** [comment]
Starting from 8.2.0.alpha1 there's a `Binder::removeBinding` that accepts either `Binding`, `HasValue` or `String propertyname`.

The `Binder::getBinding` could be made to work with a `HasValue` as well, to make it easier to find the correct `Binding`.

**Comment by pwilkin** [comment]
Yes, I'm just writing my concerns in that discussion :> 

Still, the `discard`/`commit`/`addValueChangeListener` logic for `Bindings` is a must (especially the last one, since the current version requires really ugly hacks to make sure that the value for the listener is updated after the data source is notified - I actually had to implement a `ValueChangeListener` within the... setter method for the binding).

**Comment by Legioth** [comment]
@pwilkin Yeah, that seems like a good suggestion. You previously indicated that you'd be interested in submitting pull requests for this, which would be great. The only small gotcha is that @tsuoanttila is currently also working on restructuring `Binder` internals, so there's a small risk that you'll be stepping on each other's toes.

As a workaround right now, instead of doing `field.addValueChangeListener(someListener)`, you can do something like this (plus some unsafe casts) to ensure that a listener for a specific field is invoked only after the bean has been updated:
```
binder.addValueChangeListener(event -> {
  if (field.equals(event.getSource()) {
    listener.valueChange(event);
  }
});
```

**Comment by pwilkin** [comment]
Just exchanged some remarks with @tsuoanttila on #10002 :) I will try to create a pull request with the proposed changes once his changes to Binder are merged.

**Comment by tsuoanttila** [comment]
@pwilkin is there still something missing from the expanded binding functionality this ticket is about?

**Comment by pwilkin** [comment]
@tsuoanttila Yes, I believe at least adding the `valueChangeListener` to `Binding` is needed - otherwise you have to work with the very cumbersome and unintuitive workaround for `Binder`. 

As soon as I have some free time, I'll submit the proposal.

**Comment by tsuoanttila** [comment]
I was planning to make the change that would actually not remove and re-add the value change listener from Binding all the time, but actually keep the one and same from the moment of creating the binding. This would make the listeners be called always in the same order. This can at least be done without any further API changes.

With that change you can count on the binding ValueChangeListener to be always called before any custom ones added later.

**Comment by pwilkin** [comment]
But that's still counterintuitive, still breaks if you remove and readd the listeners for some reason and still requires you to split semantic operations between the `HasValue` and the `Binding`. 

Also, the `discard()` sugar would be quite useful as well. The new API is basically still not very friendly for users who have a data-driven form instead of a user-driven form (i.e. one which strongly relies on an the form bound to an underlying data structure that might change independently of user input).

**Comment by tsuoanttila** [comment]
One quick question about the `discard()`, how should that work? I added the `read(BEAN)` that can work as a discard when you use the original bean to reset the values. Of course it won't work that well in the case of using `setBean`since the change (given that it passed validation) is immediately written to the bean. When the validation fails, then it's nice to have reset button (right now `binding.read(binder.getBean())`). For the `readBean` case, you'd need to provide the bean anyway, since it's not stored in the Binder itself.

**Comment by pwilkin** [comment]
As I said, it would be syntactic sugar, just like you mentioned (`binding.read(binder.getBean())`)

**Comment by elmot** [comment]
@pwilkin, it looks like the major part of your proposal is implemented, I will close the ticket.
Your proposal of  `discard()` method might lead to a long discussion and it's mostly up to developers taste. If you thing the method is still missing, please open another ticket, and the discussion might be continued apart of the implemented stuff.
